### ISSUE TITLE
Linking error in kwin when qt5-base was compiled with Clang/Mold

### ISSUE THREAD

**Issue by ms178** [rui314/mold#1011]
Compiling qt5-base and other qt-subprojects itself went fine with Clang-17/Mold (67eb3069b3e7c77632c3c015a3f1324b2bab87e9), but I see the following errors when compiling KDE's Kwin with Clang-17/Mold and GCC/BFD which I don't get when qt5-base and other qt-subprojects where compiled with GCC/BFD.

With Clang/Mold:

```mold: fatal: LLVM gold plugin: linking module flags 'ProfileSummary': IDs have conflicting values in '/usr/lib/libQt5XkbCommonSupport.a.llvm.1260.qxkbcommon.cpp' and 'ld-temp.o'```

With GCC-12.2.1 and BFD:

```
[ 89%] Linking CXX shared library ../bin/libkwin.so
/usr/bin/ld: /usr/lib/libQt5XkbCommonSupport.a: error adding symbols: file format not recognized
```

So I guess the problem originates from the Clang/Mold-compiled qt5-base package?! Thanks in advance for taking a look.

qt5-base: 
[PKGBUILD.txt](https://github.com/rui314/mold/files/10925081/PKGBUILD.txt)

kwin:
[PKGBUILD.txt](https://github.com/rui314/mold/files/10925083/PKGBUILD.txt)

My CFLAGS:

```
export CC=clang
export CXX=clang++
export CC_LD=mold
export CXX_LD=mold
export AR=llvm-ar
export NM=llvm-nm
export STRIP=llvm-strip
export OBJCOPY=llvm-objcopy
export OBJDUMP=llvm-objdump
export READELF=llvm-readelf
export RANLIB=llvm-ranlib
export HOSTCC=clang
export HOSTCXX=clang++
export HOSTAR=llvm-ar
export CPPFLAGS="-D_FORTIFY_SOURCE=0"
export CFLAGS="-Ofast -march=native -mtune=native -maes -mllvm -inline-threshold=1000 -mllvm -extra-vectorizer-passes -mllvm -enable-cond-stores-vec -mllvm -slp-vectorize-hor-store -mllvm -enable-loopinterchange -mllvm -enable-loop-distribute -mllvm -enable-unroll-and-jam -mllvm -enable-loop-flatten -mllvm -interleave-small-loop-scalar-reduction -mllvm -unroll-runtime-multi-exit -mllvm -aggressive-ext-opt -mllvm -enable-interleaved-mem-accesses -mllvm -enable-masked-interleaved-mem-accesses -fno-math-errno -fno-trapping-math -falign-functions=32 -funroll-loops -fno-semantic-interposition -fcf-protection=none -mharden-sls=none -fomit-frame-pointer -mprefer-vector-width=256 -flto -fgnuc-version=6.5.0"
export CXXFLAGS="${CFLAGS}"
export LDFLAGS="-Wl,--lto-O3,-O3,-Bsymbolic-functions,--as-needed -mllvm -extra-vectorizer-passes -mllvm -enable-cond-stores-vec -mllvm -slp-vectorize-hor-store -mllvm -enable-loopinterchange -mllvm -enable-loop-distribute -mllvm -enable-unroll-and-jam -mllvm -enable-loop-flatten -mllvm -interleave-small-loop-scalar-reduction -mllvm -unroll-runtime-multi-exit -mllvm -aggressive-ext-opt -mllvm -enable-interleaved-mem-accesses -mllvm -enable-masked-interleaved-mem-accesses -march=native -maes -fuse-ld=mold -flto -Wl,-zmax-page-size=0x200000"
CCLDFLAGS="$LDFLAGS"
CXXLDFLAGS="$LDFLAGS"
export ASFLAGS="-D__AVX__=1 -D__AVX2__=1 -msse2avx -D__FMA__=1"
```

**Comment by ishitatsuyuki** [comment]
Maybe this has something to do with you passing a lot of compiler fine tuning flags? "IDs have conflicting values in" sounds like the LTO optimizer complaning about bitcode compiled with different flags, while "error adding symbols: file format not recognized" sounds like the BFD linker not recognizing the bitcode at all (did you tell it to load the LLVM LTO plugin?).

**Comment by ms178** [comment]
Hm, I don't know if it is related, but when compiling LLVM-git with mold, I see the following warning when compiling compiler-rt and openmp as runtime which hints at some mismatch which could indeed be a problem in the issue reported here as well: 

`mold: warning: LLVM gold plugin: Linking two modules of different target triples: '/tmp/makepkg/llvm-git/src/llvm-project/build/lib/libLLVMSupport.a.llvm.7436846.Threading.cpp' is 'x86_64-unknown-linux-gnu' whereas 'ld-temp.o' is 'x86_64-pc-linux-gnu'`

**Comment by ishitatsuyuki** [comment]
I guess, if you can add/remove one flag at a time and figure out which one is causing the problem, that might make this issue much easier to diagnose.

**Comment by ms178** [comment]
I think I found the culprit, qt5-base disregards my makepkg.conf settings and uses BFD instead of mold. I need to find a way to force the package to accept mold.

**Comment by ms178** [comment]
Good news, I got it to work finally. I needed to patch the /mkspecs/common/clang.conf file for using Mold with Clang on qt5-base; the default Arch patches only touch the corresponding GCC file. I just had to make sure to also use flto=thin for Kwin from now on (I used full lto there before) but the build there succeeds now finally.

I've uploaded the relevant files in my repo: https://github.com/ms178/archpkgbuilds/commit/60e325129307e5a6958396cc01fd254ec5f3eed7

They are a bit hacky, but do the trick. There is probably not something on the Mold side to ease this; I hope that with the CMAKE build system in qt6 onwards, this problematic build system handling will go away.
